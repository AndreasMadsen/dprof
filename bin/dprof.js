#!/bin/env node

var endpoint = require('endpoint');
var extend = require('util-extend');

process.stdin.pipe(endpoint(function (err, dprof) {
  if (err) throw err;
  dprof = JSON.parse(dprof);

  var content = new Node(dprof.root, dprof.total, 1, new State(), new Root());

  var svg = new Element('svg').attr({ xmlns: 'http://www.w3.org/2000/svg', width: 1000 });

  process.stdout.write(svg.start() + '\n');
  process.stdout.write(content.stringify() + '\n');
  process.stdout.write(svg.end());
}));

function Root() {
  this.top = 0;
}

function State() {
  this.y = 0;
}

function Node(node, total, depth, state, parent) {
  this.name = node.name;
  this.init = Math.floor(node.init / total * 1000);
  this.before = Math.floor((node.init + node.before) / total * 1000);
  this.after = Math.floor((node.init + node.after) / total * 1000);
  this.stack = node.stack;

  this.depth = depth;
  this.state = state;
  this.parent = parent;
  this.top = state.y;

  this.elems = [];
  this.elems.push(new Element('rect').attr({
    x: this.init,
    y: this.top,
    width: this.before - this.init,
    height: 20,
    fill: 'SteelBlue',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('rect').attr({
    x: this.before,
    y: this.top,
    width: this.after - this.before,
    height: 20,
    fill: 'IndianRed',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('line').attr({
    x1: this.init,
    y1: this.parent.top + 20,
    x2: this.init,
    y2: this.top + 20,
    style: 'stroke: #000; stroke-width: 2px'
  }));

  this.state.y += 20;
  this.children = node.children.map(function (child) {
    return new Node(child, total, depth + 1, state, this);
  }, this);
}

Node.prototype._indent = function () {
  var s = '';
  for (var i = 0; i < this.depth; i++) {
    s += '  ';
  }
  return s;
};

Node.prototype.stringify = function () {
  var content = [];
  this.elems.forEach(function (elem) {
    content.push(this._indent() + elem.singleton());
  }, this);
  this.children.forEach(function (child) {
    content.push(child.stringify());
  }, this);

  return content.join('\n');
};

function Element(name) {
  this._name = name;
  this._attr = {};
}

Element.prototype.attr = function (attr) {
  extend(this._attr, attr);
  return this;
};

Element.prototype.start = function () {
  var start = '<' + this._name;
  Object.keys(this._attr).forEach(function (k) {
    start += ' ' + k + '="' + this._attr[k] + '"';
  }, this);
  start += '>';
  return start;
};

Element.prototype.end = function () {
  return '</' + this._name + '>';
};

Element.prototype.singleton = function () {
  return this.start() + this.end();
};
