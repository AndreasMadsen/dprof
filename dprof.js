'use strict';

const providers = process.binding('async_wrap').Providers;
const asyncWrap = require('./async_wrap.js');
const zlib = require('zlib');
const fs = require('fs');

const version = require('./package.json').version;
const processStart = process.hrtime();

// Change the default stackTraceLimit if it haven't allready been overwritten
if (process.execArgv.indexOf('--stack_trace_limit') === -1 && Error.stackTraceLimit === 10) {
  Error.stackTraceLimit = 8;
}

//
// Define node class
//

function Site(site) {
  this.description = site.toString();
  this.filename = site.getFileName();
  this.line = site.getLineNumber();
  this.collum = site.getColumnNumber();
}

function timestamp() {
  const t = process.hrtime(processStart);
  return t[0] * 1e9 + t[1];
}

function Node(name, stack) {
  this.name = name;
  this._init = timestamp();
  this._before = [];
  this._after = [];
  this.children = [];
  this.stack = stack.map(function (site) {
    return new Site(site);
  });
}

Node.prototype.add = function (handle) {
  const node = new Node(handle.constructor.name, asyncWrap.stackTrace(3));
  this.children.push(node);
  return node;
};

Node.prototype.before = function () {
  this._before.push(timestamp());
};

Node.prototype.after = function () {
  this._after.push(timestamp());
};

Node.prototype.toJSON = function () {
  return {
    name: this.name,
    init: this._init,
    before: this._before,
    after: this._after,
    stack: this.stack,
    children: this.children
  };
};

Node.prototype.rootIntialize = function () {
  this._init = 0;
  this._before.push(0);
};

//
// Setup hooks
//

asyncWrap.setup(asyncInit, asyncBefore, asyncAfter);

const root = new Node('root', asyncWrap.stackTrace(2));
      root.rootIntialize();
let state = root;

function asyncInit(provider, parent) {
  if (provider === providers.TIMERWRAP) {
    this._dprofIgnore = true;
    return;
  }

  if (parent) {
    this._dprofState = parent._dprofState.add(this, provider, parent);
  } else {
    this._dprofState = state.add(this, provider, parent);
  }
}

function asyncBefore() {
  if (this._dprofIgnore) return;

  this._dprofState.before();
  state = this._dprofState;
}

function asyncAfter() {
  if (this._dprofIgnore) return;
  
  this._dprofState.after();
  state = root;
}

// The root job is done when process.nextTick is called
asyncWrap.disable();
process.nextTick(function () {
  root.after();
});
asyncWrap.enable();

//
// Print result
//

process.on('exit', function () {
  // even though zlib is sync, it still fires async_wrap events,
  // so disable asyncWrap just to be sure.
  asyncWrap.disable();

  const data = {
    'total': timestamp(),
    'version': version,
    'root': root
  };

  if (process.env.NODE_DPROF_DEBUG) {
    fs.writeFileSync('./dprof.json', JSON.stringify(data, null, 1));
  } else {
    fs.writeFileSync('./dprof.json.gz', zlib.gzipSync(JSON.stringify(data)));
  }
});
