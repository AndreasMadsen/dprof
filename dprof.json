{
 "name": "root",
 "init": 0,
 "before": 0,
 "after": 4319266,
 "children": [
  {
   "name": "Timer",
   "init": 2044586,
   "before": 548667,
   "after": 1409916,
   "children": [
    {
     "name": "FSReqWrap",
     "init": 3039113,
     "before": 458225,
     "after": 788615,
     "children": [
      {
       "name": "FSReqWrap",
       "init": 3776193,
       "before": 164991,
       "after": 213630,
       "children": []
      },
      {
       "name": "FSReqWrap",
       "init": 3803875,
       "before": 188243,
       "after": 268189,
       "children": [
        {
         "name": "FSReqWrap",
         "init": 4063866,
         "before": 39850,
         "after": 74510,
         "children": []
        }
       ]
      }
     ]
    },
    {
     "name": "FSReqWrap",
     "init": 3279865,
     "before": 623396,
     "after": 650968,
     "children": []
    }
   ]
  }
 ]
}