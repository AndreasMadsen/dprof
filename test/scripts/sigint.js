'use strict';

// Should fail with either:
// process.argv.push('--dprof-no-sigint');
// process.env.DPROF_NO_SIGINT = '';
require('../../dprof.js');

setInterval(function () {}, 1000);
setTimeout(function () {
  process.kill(process.pid, 'SIGINT');
}, 200);
