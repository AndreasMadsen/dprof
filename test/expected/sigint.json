{
 "stdout": "",
 "stderr": "init { uid: 3, type: 'SIGNALWRAP', triggerId: 1 }\ninit { uid: 4, type: 'Timeout', triggerId: 1 }\ninit { uid: 5, type: 'TIMERWRAP', triggerId: 1 }\ninit { uid: 6, type: 'Timeout', triggerId: 1 }\ninit { uid: 7, type: 'TIMERWRAP', triggerId: 1 }\nbefore { uid: 7 }\nbefore { uid: 6 }\nafter { uid: 6 }\nafter { uid: 7 }\ndestroy { uid: 6 }\nbefore { uid: 3 }\n",
 "dump": {
  "total": 217401553,
  "version": "1.0.0",
  "root": {
   "name": "root",
   "parent": null,
   "uid": 1,
   "init": 0,
   "destroy": 11341900,
   "before": [
    0
   ],
   "after": [
    11317122
   ],
   "unref": [],
   "ref": [],
   "initRef": true,
   "stack": [
    {
     "description": "Module._compile (module.js:569:30)",
     "filename": "module.js",
     "line": 569,
     "collum": 30
    },
    {
     "description": "Object.Module._extensions..js (module.js:580:10)",
     "filename": "module.js",
     "line": 580,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:503:32)",
     "filename": "module.js",
     "line": 503,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:466:12)",
     "filename": "module.js",
     "line": 466,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:458:3)",
     "filename": "module.js",
     "line": 458,
     "collum": 3
    },
    {
     "description": "Module.require (module.js:513:17)",
     "filename": "module.js",
     "line": 513,
     "collum": 17
    },
    {
     "description": "require (internal/module.js:11:18)",
     "filename": "internal/module.js",
     "line": 11,
     "collum": 18
    },
    {
     "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/sigint.js:6:1)",
     "filename": "/test/scripts/sigint.js",
     "line": 6,
     "collum": 1
    },
    {
     "description": "Module._compile (module.js:569:30)",
     "filename": "module.js",
     "line": 569,
     "collum": 30
    },
    {
     "description": "Object.Module._extensions..js (module.js:580:10)",
     "filename": "module.js",
     "line": 580,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:503:32)",
     "filename": "module.js",
     "line": 503,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:466:12)",
     "filename": "module.js",
     "line": 466,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:458:3)",
     "filename": "module.js",
     "line": 458,
     "collum": 3
    },
    {
     "description": "Function.Module.runMain (module.js:605:10)",
     "filename": "module.js",
     "line": 605,
     "collum": 10
    },
    {
     "description": "startup (bootstrap_node.js:158:16)",
     "filename": "bootstrap_node.js",
     "line": 158,
     "collum": 16
    },
    {
     "description": "bootstrap_node.js:575:3",
     "filename": "bootstrap_node.js",
     "line": 575,
     "collum": 3
    }
   ],
   "children": [
    3,
    4,
    5,
    6,
    7
   ]
  },
  "nodes": [
   {
    "name": "SIGNALWRAP",
    "parent": 1,
    "uid": 3,
    "init": 5923476,
    "destroy": null,
    "before": [
     217041102
    ],
    "after": [],
    "unref": [
     6645007
    ],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "Signal.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "process.<anonymous> (internal/process.js:203:20)",
      "filename": "internal/process.js",
      "line": 203,
      "collum": 20
     },
     {
      "description": "emitTwo (events.js:125:13)",
      "filename": "events.js",
      "line": 125,
      "collum": 13
     },
     {
      "description": "process.emit (events.js:213:7)",
      "filename": "events.js",
      "line": 213,
      "collum": 7
     },
     {
      "description": "_addListener (events.js:248:14)",
      "filename": "events.js",
      "line": 248,
      "collum": 14
     },
     {
      "description": "process.addListener (events.js:298:10)",
      "filename": "events.js",
      "line": 298,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/dprof.js:206:11)",
      "filename": "/dprof.js",
      "line": 206,
      "collum": 11
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Module.require (module.js:513:17)",
      "filename": "module.js",
      "line": 513,
      "collum": 17
     },
     {
      "description": "require (internal/module.js:11:18)",
      "filename": "internal/module.js",
      "line": 11,
      "collum": 18
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/sigint.js:6:1)",
      "filename": "/test/scripts/sigint.js",
      "line": 6,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "Timeout",
    "parent": 1,
    "uid": 4,
    "init": 8822396,
    "destroy": null,
    "before": [],
    "after": [],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "emitInitS (async_hooks.js:327:7)",
      "filename": "async_hooks.js",
      "line": 327,
      "collum": 7
     },
     {
      "description": "new Timeout (timers.js:592:5)",
      "filename": "timers.js",
      "line": 592,
      "collum": 5
     },
     {
      "description": "createRepeatTimeout (timers.js:561:15)",
      "filename": "timers.js",
      "line": 561,
      "collum": 15
     },
     {
      "description": "exports.setInterval (timers.js:553:10)",
      "filename": "timers.js",
      "line": 553,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/sigint.js:8:1)",
      "filename": "/test/scripts/sigint.js",
      "line": 8,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "TIMERWRAP",
    "parent": 1,
    "uid": 5,
    "init": 9475984,
    "destroy": null,
    "before": [],
    "after": [],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "Timer.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "new TimersList (timers.js:218:17)",
      "filename": "timers.js",
      "line": 218,
      "collum": 17
     },
     {
      "description": "createTimersList (timers.js:203:16)",
      "filename": "timers.js",
      "line": 203,
      "collum": 16
     },
     {
      "description": "insert (timers.js:185:27)",
      "filename": "timers.js",
      "line": 185,
      "collum": 27
     },
     {
      "description": "exports.active (timers.js:142:3)",
      "filename": "timers.js",
      "line": 142,
      "collum": 3
     },
     {
      "description": "createRepeatTimeout (timers.js:566:3)",
      "filename": "timers.js",
      "line": 566,
      "collum": 3
     },
     {
      "description": "exports.setInterval (timers.js:553:10)",
      "filename": "timers.js",
      "line": 553,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/sigint.js:8:1)",
      "filename": "/test/scripts/sigint.js",
      "line": 8,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "Timeout",
    "parent": 1,
    "uid": 6,
    "init": 10063463,
    "destroy": 216929715,
    "before": [
     216101042
    ],
    "after": [
     216674453
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "emitInitS (async_hooks.js:327:7)",
      "filename": "async_hooks.js",
      "line": 327,
      "collum": 7
     },
     {
      "description": "new Timeout (timers.js:592:5)",
      "filename": "timers.js",
      "line": 592,
      "collum": 5
     },
     {
      "description": "createSingleTimeout (timers.js:472:15)",
      "filename": "timers.js",
      "line": 472,
      "collum": 15
     },
     {
      "description": "setTimeout (timers.js:456:10)",
      "filename": "timers.js",
      "line": 456,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/sigint.js:9:1)",
      "filename": "/test/scripts/sigint.js",
      "line": 9,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "TIMERWRAP",
    "parent": 1,
    "uid": 7,
    "init": 10310313,
    "destroy": null,
    "before": [
     215366922
    ],
    "after": [
     216819400
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "Timer.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "new TimersList (timers.js:218:17)",
      "filename": "timers.js",
      "line": 218,
      "collum": 17
     },
     {
      "description": "createTimersList (timers.js:203:16)",
      "filename": "timers.js",
      "line": 203,
      "collum": 16
     },
     {
      "description": "insert (timers.js:185:27)",
      "filename": "timers.js",
      "line": 185,
      "collum": 27
     },
     {
      "description": "exports.active (timers.js:142:3)",
      "filename": "timers.js",
      "line": 142,
      "collum": 3
     },
     {
      "description": "createSingleTimeout (timers.js:476:3)",
      "filename": "timers.js",
      "line": 476,
      "collum": 3
     },
     {
      "description": "setTimeout (timers.js:456:10)",
      "filename": "timers.js",
      "line": 456,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/sigint.js:9:1)",
      "filename": "/test/scripts/sigint.js",
      "line": 9,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   }
  ]
 }
}